packages/client/src/compatibility.ts

```diff
@@ -223,28 +229,65 @@ export const minCompatLevel = (levels: {
 }) => Math.min(levels.args, levels.values)
 
 const withOptionalToken =
-  <T, D>(
+  <T, D, A extends [...any[], CompatibilityToken]>(
     compatibilityToken: Promise<CompatibilityToken<D>>,
-    fn: (runtime: CompatibilityToken) => T,
-  ): WithOptionalRuntime<T, D> =>
-  (runtime?: CompatibilityToken): any =>
-    runtime ? fn(runtime) : compatibilityToken.then(fn)
+    fn: (...args: A) => T,
+  ): WithOptionalRuntime<T, D, A extends [...infer R, any] ? R : []> =>
+  (...args: any): any => {
+    const lastElement = args.at(-1)
+    if (lastElement instanceof CompatibilityToken) {
+      return fn(...args)
+    }
+    return compatibilityToken.then((token) => (fn as any)(...args, token))
+  }
 
-export type WithOptionalRuntime<T, D> = {
+export type WithOptionalRuntime<T, D, A extends any[]> = {
   /**
    * Returns the result after waiting for the runtime to load.
    */
-  (): Promise<T>
+  (...args: A): Promise<T>
   /**
    * Returns the result synchronously with the loaded runtime.
    */
-  (runtime: CompatibilityToken<D>): T
+  (...args: [...A, runtime: CompatibilityToken<D>]): T
 }
 
 export interface CompatibilityFunctions<D> {
   /**
    * Returns the `CompatibilityLevel` for this call comparing the descriptors
    * generated on dev time with the current live metadata.
    */
-  getCompatibilityLevel: WithOptionalRuntime<CompatibilityLevel, D>
+  getCompatibilityLevel(): Promise<CompatibilityLevel>
```

Comment:
Even though `: WithOptionalRuntime<...>` would work, this is part of the public API and the docs the IDE shows are of the `WithOptionalRuntime` overloads.

So I've decided to inline them instead

Reply:
I was about to say it is a great idea to inline them. Nice!